libraries/AP_Compass/AP_Compass_AK8963.cpp

```diff
@@ -106,36 +106,21 @@
 
 extern const AP_HAL::HAL& hal;
 
-AP_Compass_AK8963::AP_Compass_AK8963(Compass &compass) :
+AP_Compass_AK8963::AP_Compass_AK8963(Compass &compass, AP_AK8963_SerialBus *bus) :
     AP_Compass_Backend(compass),
     _state(STATE_UNKNOWN),
     _initialized(false),
     _last_update_timestamp(0),
-    _last_accum_time(0)
+    _last_accum_time(0),
+    _bus(bus)
 {
     _mag_x_accum =_mag_y_accum = _mag_z_accum = 0;
     _mag_x =_mag_y = _mag_z = 0;
     _accum_count = 0;
     _magnetometer_adc_resolution = AK8963_16BIT_ADC;
+    init();
 }
 
-AP_Compass_Backend *AP_Compass_AK8963::detect(Compass &compass)
-{
-    AP_Compass_AK8963 *sensor = new AP_Compass_AK8963(compass);
-
-    if (sensor == nullptr) {
-        return nullptr;
-    }
-
-    if (!sensor->init()) {
-        delete sensor;
-        return nullptr;
-    }
-
-    return sensor;
-}
```

Comment:
Discussed with Tridge, I'll change the implementation of the generic bus by adding detect methods that will be specific to the bus like detect_i2c1(I2C_ADDR);
It will be the same for the IMUs